### PR TITLE
[v2] Add pre-commit and CI static checks

### PR DESCRIPTION
**Pull request by turbaszek**
Code quality is important, especially in an open source project where the code will be maintained for a long time and by multiple people. So I would like to suggest using https://pre-commit.com framework to automate a lot of stuff and assure that our code is nice. 

It plays nicely with Github action and allows a plethora of checks (building TOC, license insert, liniting etc) and gives possibility to build custom ones. For example here's configuration from Apache Airflow: 
https://github.com/apache/airflow/blob/master/.pre-commit-config.yaml

### Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO)
- [ ] Tests have been added
- [ ] A PR is opened to update the documentation on https://github.com/kedacore/keda-docs
- [ ] Changelog has been updated

### CODE REVIEW

File: .pre-commit-config.yaml
Comment (tomkerkhove) [reply .pre-commit-config.yaml:29]: If this gets merged, please don't forget to open issues for these.

File: .github/workflows/static-checks.yml
Comment (tomkerkhove) [inline .github/workflows/static-checks.yml:1]:
```suggestion
name: Code Quality
```

**Review by tomkerkhove — COMMENTED** [review]
This looks OK to me, thanks!

Just make sure to target v2 instead of master so that @zroubalik doesn't have to plow through all of these.

Just a tip, let's discuss this in an issue first in the future because if we didn't like this then you would've wasted a ton of work

**Comment by turbaszek** [comment]
@tomkerkhove thanks for the review! Regarding:

> Just a tip, let's discuss this in an issue first in the future because if we didn't like this then you would've wasted a ton of work

I just copy pasted pre-commit config from my other go repo so not a lot of work 😉

**Comment by turbaszek** [comment]
Oh, DCO doesn't accept PR added from GitHub review? :,

**Comment by turbaszek** [comment]
@zroubalik I've rebased this change onto v2 👍

File: .pre-commit-config.yaml
Comment (turbaszek) [inline .pre-commit-config.yaml:68]: Inspired by @zroubalik in https://github.com/kedacore/keda/pull/1026#discussion_r475599632
```
Check if scalers are sorted in scaler_handler.go............................Failed
hookid: sort-scalers

Scalers are not sorted alphabetical in pkg/scaling/scale_handler.go
```

**Comment by mik-laj** [comment]
<3 😻

File: CONTRIBUTING.md
Comment (mik-laj) [inline CONTRIBUTING.md:95]:
```suggestion
The project uses many static checks using [pre-commit](https://pre-commit.com/). Every change is checked on CI and if it does not pass the tests it cannot be accepted. If you want to check locally then you should install Python3.6 or newer together and run:
```bash
pip install pre-commit
# or
brew install pre-commit
```
then from project directory run `pre-commit install`. For more installation options
visit the [pre-commits](https://pre-commit.com).

To turn on pre-commit checks for commit operations in git, run:
```bash
pre-commit install
```

To run all checks on your staged files, run:
```bash
pre-commit run
```

To run all checks on all files, run:
```bash
pre-commit run --all-files
```

Pre-commit check results are also attached to your PR through integration with CI.
```

Comment (turbaszek) [reply CONTRIBUTING.md:95]: Looks good! Thanks @mik-laj 🐈

File: CONTRIBUTING.md
Comment (tomkerkhove) [inline CONTRIBUTING.md:7]: That's great, but how does this work? Should add some docs on this :)

File: CONTRIBUTING.md
Comment (tomkerkhove) [inline CONTRIBUTING.md:16]: Nit - Align casing with other titles

**Review by tomkerkhove — COMMENTED** [review]
LGTM but I'll leave it up to @zroubalik to give the sign-off

File: CONTRIBUTING.md
Comment (turbaszek) [reply CONTRIBUTING.md:7]: I would say that's enough to google `doctoc` to find out more information. Also not sure where to put such docs

File: .pre-commit-config.yaml
Comment (potiuk) [reply .pre-commit-config.yaml:77]: Love it :heart:

File: CONTRIBUTING.md
Comment (potiuk) [reply CONTRIBUTING.md:7]: Yeah. And wit pre-commits you basically do not have to worry about it - just install pre-commit and you forget about it..

**Review by potiuk — APPROVED** [review]

File: CONTRIBUTING.md
Comment (tomkerkhove) [reply CONTRIBUTING.md:7]: That's a bit of an odd argument as that would mean we don't have to doc the pre-commit thingy but it's ok.

> Yeah. And wit pre-commits you basically do not have to worry about it - just install pre-commit and you forget about it..

We tend to make it easy for folks to contribute so you cannot assume everybody has `pre-commit` installed, that's why we have a CI.

Comment (turbaszek) [reply CONTRIBUTING.md:7]: > We tend to make it easy for folks to contribute so you cannot assume everybody has pre-commit installed, that's why we have a CI.

As we have totally lightweight pre-commit setup we should probably encourage people to use them. It's good to teach people how to automate stuff and focus on the most important things like the code itself and documentation. However, that's only my personal opinion.

Comment (tomkerkhove) [reply CONTRIBUTING.md:7]: That's definitely the case and I agree, but we can't expect everyone to use it.

Comment (mik-laj) [reply CONTRIBUTING.md:7]: We also have it integrated on the CI, so if someone does not use pre-commit locally, they will get a readable diff on the CI that will show what changes should be made.

Comment (tomkerkhove) [reply CONTRIBUTING.md:7]: Yes, that was my point on https://github.com/kedacore/keda/pull/1025#discussion_r476294569 indeed

Comment (potiuk) [reply CONTRIBUTING.md:7]: Yeah. Pre-commits are working like a charm on Apache Airflow for ? year now and we have plenty of pre-commits running/ You can see it working here for example: https://github.com/apache/airflow/runs/1026196376?check_suite_focus=true#step:7:108 - it integrates very well with CI. you simply run `pre-commit run --all-files` in the CI and you get the very same checks executed in CI as locally. 

And as @mik-laj mentioned - with `--show-diff-on-failure` flag it will even show you what modification would be made if you run pre-commit locally. In case it fails in CI it will even tell you what to do 'pip install pre-commit' and then `pre-commit run <id of pre-commit that failed>`.  But you still can fix it manually if you prefer (though a lot of pre-commit checks can fix things for you automatically (like end-of-file checking, sorting import in python etc.) and then the only thing you have to do you have to commit modified files. And you have literally hundreds of hooks that you can start using. 

So you do not have to even install it as pre-commit (though it is super-easy `pre-commit install` will do the job after installit it with pip. The nice thing about it is that you simply forget about it once you install - it will automatically run all the precommits that were added by others since and you will actually fix the problems even before others see them.

I do not see any drawbacks of using pre-commit to be honest  if your project's goal is to keep some consistency, it's the perfect tool for the job IMHO.

File: README.md
Comment (ahmelsayed) [inline README.md:158]: fix git merge conflict markers?

**Review by TobKed — APPROVED** [review]

File: README.md
Comment (turbaszek) [reply README.md:158]: Oh, will add check for those 😄

Comment (turbaszek) [reply README.md]: Thanks!

**Comment by turbaszek** [comment]
> I'm assuming the CI pre-commit will fail if something needs to be fixed, right? I haven't used pre-commit before but looks neat.

That's right! For example:
```
Check if scalers are sorted in scaler_handler.go............................Failed
hookid: sort-scalers

Scalers are not sorted alphabetical in pkg/scaling/scale_handler.go
```

Most of pre-commit hooks we use apply the changes automatically during `git commit` (remove white spaces, go-fmt, go-lint) so not to much to do for developers

**Review by mik-laj — APPROVED** [review]

**Comment by turbaszek** [comment]
@tomkerkhove @zroubalik are we good to go?

File: CONTRIBUTING.md
Comment (tomkerkhove) [reply CONTRIBUTING.md:7]: Don't get me wrong, I'm not saying it's a bad thing or so; I was just saying that it's fine and the CI will cover it but we shouldn't force people to use it.

I'm not saying you are, it was just a general remark :) This escalated quickly.

File: CONTRIBUTING.md
Comment (tomkerkhove) [inline CONTRIBUTING.md:87]: Let's emphasize it's not a must

```suggestion
This project is using [pre-commits](https://pre-commit.com) to ensure the quality of the code, however, this is not a must to contribute.

  Every change
```

**Review by tomkerkhove — APPROVED** [review]
LGTM, just one more change.

File: CONTRIBUTING.md
Comment (turbaszek) [reply CONTRIBUTING.md:87]: How about:
> We encourage you to use pre-commits but it's not a must. 

Imho same meaning but has a nice call to action. Believe me, once you start using pre-commits you you will appreciate them 😄

Comment (tomkerkhove) [reply CONTRIBUTING.md:87]: I believe you and I use it myself, but we should not force people to do so.

I'll leave it up to @zroubalik to decide.

Comment (potiuk) [reply CONTRIBUTING.md:87]: Yeah. I think "Encourage" is a good word.

In Apache Airflow this often works like that:

a) contributor submit CI build several times and it fails for different reason
b) the contributor fixes it manually and re-submits (and it fails again as it was not perfect fix)
c) one of the committers gently suggests "we encourage you to use pre-commits" to save your time (and points out to the dos) 
d) the contributor comes back with green PR with "I wish I knew pre-commit before" comment ...
e) the contributor starts advising other contributors to use pre-commit

Comment (tomkerkhove) [reply CONTRIBUTING.md:87]: Thanks for the suggestion.

Comment (zroubalik) [reply CONTRIBUTING.md:87]: Yeah I'd definitely add a note over there as @tomkerkhove suggested. We shouldn't force that or gave users such impression. I'd add a note something like this `We encourage you to use pre-commits but it's not a must to contribute.`

**Review by zroubalik — APPROVED** [review]
LGTM, just minor remark on "enforcing" pre-commit. Thanks for the contribution

File: CONTRIBUTING.md
Comment (turbaszek) [reply CONTRIBUTING.md]: Added

**Review by zroubalik — APPROVED** [review]
LGTM

**Comment by mik-laj** [comment]
:tada:

**Comment by turbaszek** [comment]
Thanks all! 🚀

**Comment by potiuk** [comment]
:heart:

**Comment by kaxil** [comment]
<3

**Pull request merged**
